Add piston example for tensor_mechanics module

GregVernon opened the pull request:
## Reason
Suggested to add from this GitHub Discussion: https://github.com/idaholab/moose/discussions/18531
Where at least one new user was following an example on YouTube, and found an old syntax issue.
Adding the example to MOOSE to keep it up-to-date to help more new users.

## Design
Adds a working input file and input exodus file for example.

## Impact
Supports new users of Tensor Mechanics module

Commit:
Add piston example for tensor_mechanics module

Closes https://github.com/idaholab/moose/issues/18619

Suggested to add from this GitHub Discussion:
https://github.com/idaholab/moose/discussions/18531

GregVernon commented:
@hugary1995

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,178 @@
+[]
+
+# This is where mesh adaptivity magic happens:
+[./Adaptivity]
+  steps = 1
+  #initial_steps = 2
+  max_h_level = 3
+  cycles_per_step = 1
+  initial_marker = uniform
+  marker = errorFraction
+  [./Markers]
+    [./uniform]
+      type = UniformMarker
+      mark = refine
+    [../]
+    
+    [./errorFraction]
+      type = ErrorFractionMarker
+      coarsen = 0.5
+      indicator = gradientJump
+      refine = 0.5
+    [] 
+ [../]
+
+  [./Indicators]
+    [./gradientJump]
+      type = GradientJumpIndicator
+      variable = disp_y
+    []
+  []
+[../]
```
I think we still have some issues with using adaptivity on mechanics problems. 

If you want to use adaptivity, the safest way is to use total small deformation that does not need stateful properties.

GregVernon replied:
I would like to use adaptivity in this example, can you suggest an adaptivity block?

jiangwen84 replied:
In the TensorMechanicsMasterAction, use `incremental = false` and `strain=SMALL`. Then it will not use any stateful properties.

GregVernon replied:
How's this https://github.com/idaholab/moose/pull/18543/commits/bad294ccccf393bf9d440f0003ebfa4b6a8262e8?

dschwen replied:
could you run `hit format` on the input?

GregVernon replied:
@dschwen - see https://github.com/idaholab/moose/pull/18543/commits/72d4a16b0e5f22bb204e73df38d448bab22588d9

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,178 @@
+  [../]
+[]
+
+[Kernels]
+  [./TensorMechanics]
+    #Stress divergence kernels
+    displacements = 'disp_x disp_y disp_z'
+  [../]
+[]
```
use tensor mechanics master action.

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/9804180a0c2c0919c5360dc81ce21a1fd8368e75

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,178 @@
+  [../]
+[]
+
+[AuxVariables]
+  [./von_mises]
+    #Dependent variable used to visualize the Von Mises stress
+    order = CONSTANT
+    family = MONOMIAL
+  [../]
+[]
+
+[AuxKernels]
+  [./von_mises_kernel]
+    #Calculates the von mises stress and assigns it to von_mises
+    type = RankTwoScalarAux
+    variable = von_mises
+    rank_two_tensor = stress
+    execute_on = timestep_end
+    scalar_type = VonMisesStress
+  [../]
+[]
```
With tensor mechanics action, you can use `generate_output = vonmises_stress` for outputting.

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/b9fba72d5aea38f9559e6c654d576ce79737a490

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,178 @@
+    [./load]
+      #Applies the pressure
+      boundary = load_surf
+      function = rampLinear
```
You can remove [Functions] block and use `function = 't*550e5'` here

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/c6d52d5c378266ab856c71b0a6f872ef94f495a9

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,178 @@
+      #Applies the pressure
+      boundary = load_surf
+      function = rampLinear
+      displacements = 'disp_x disp_y disp_z'
```
It might be convenient to define  `displacements = 'disp_x disp_y disp_z'` in `GlobalParams`

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/18eb57173e092c63d52a5c514153b2c97bb72e48

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,178 @@
+[]
+
+[Materials]
+  active = 'density_steel stress strain elasticity_tensor_steel'
```
delete this line.

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/c5a24c429e9bad68939ae1cdd388eddd1cd99c76

jiangwen84 reviewed (changes requested):
Some minor comments. 

You also need to add `tests` file to at least check the syntax.

Commit:
Remove Functions block, inline function in pressure

Commit:
Add GlobalParams block defining displacements variable

Commit:
Add TensorMechanicsMaster action

Commit:
Remove manual setup of VonMises output

Commit:
Remove unnecessary "active" line in Materials block

Commit:
Add tests file, check syntax only

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,152 @@
+  []
+[../]
+
+[Variables]
+  [./disp_x]
+    order = FIRST
+    family = LAGRANGE
+  [../]
+  [./disp_y]
+    order = FIRST
+    family = LAGRANGE
+  [../]
+  [./disp_z]
+    order = FIRST
+    family = LAGRANGE
+  [../]
+[]
```
Since you use `add_variables = true` in Master Action, you do not need to list those variables.

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/5696be02cd2e34ed99a48da6963927b6f263bb42

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,152 @@
+    type = ComputeFiniteStrainElasticStress
+    block = 1
+  [../]
+  [./density_steel]
+    #Defines the density of steel
+    type = GenericConstantMaterial
+    block = 1
+    prop_names = density
+    prop_values = 7850 # kg/m^3
+  [../]
```
do we need `density`?

GregVernon replied:
We don't _need_ it for this simulation.  I'd left it in because I'd been testing some dynamics.  I can remove.

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/e36103757a45619ae4457914d27d693054582e9a

Commit:
Remove density from material

Commit:
Remove variables block, unnecessary due to Master action

Commit:
Remove statefull properties

Finite strain -> small strain

Commit:
Remove extraneous "displacements" command

GregVernon commented:
@jiangwen84 -- would you like _me_ to "resolve" conversations, or should I leave that to you/a reviewer?

jiangwen84 commented on modules/tensor_mechanics/examples/piston/piston_params.i:
```diff
@@ -0,0 +1,130 @@
+# This is where mesh adaptivity magic happens:
+[./Adaptivity]
+  steps = 1
+  #initial_steps = 2
```
delete this line if you do not need.

GregVernon replied:
See https://github.com/idaholab/moose/pull/18543/commits/98a239e402b3cf2b87f62288a0962570948c16af

Commit:
Remove unused commands, format comments

Commit:
Output of `hit format`

Executed `~/projects/moose/framework/contrib/hit/hit format -i piston_params.i`

GregVernon commented:
@jiangwen84 -- is there anything else that needs to be done?

jiangwen84 commented:
> @jiangwen84 -- is there anything else that needs to be done?

@GregVernon it looks good to me. This PR has some pre-check errors that needs fixed. If it passes all tests, I think we can merge it.

Commit:
Output of delete_trailing_whitespace.sh

GregVernon commented:
@jiangwen84 -- the precheck is giving an error:

```
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
```

Do I need to create an issue, then specify this PR as closing it?

jiangwen84 commented:
> @jiangwen84 -- the precheck is giving an error:
> 
> ```
> ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
> ```
> 
> Do I need to create an issue, then specify this PR as closing it?

Correct. Every PR needs an issue number.

hugary1995 commented:
FYI, at least one commit message has to contain that ticket number. Editing the PR description probably isn't enough.

GregVernon commented:
> FYI, at least one commit message has to contain that ticket number. Editing the PR description probably isn't enough.

Thanks @hugary1995 -- I reworded the initial commit to include the ticket I made, `precheck` is passing now.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/797491/) on 2524e8f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18543/site)

This comment will be updated on new commits.

jiangwen84 commented:
@GregVernon  It looks great now.
